File: wrappers/kotlin/src/commonMain/kotlin/crypto/Key.kt
package askar.crypto

import askar.Askar
import askar.enums.KeyAlgs
import askar.enums.KeyMethod
import askar.enums.SigAlgs
import askar.enums.keyAlgFromString
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

class Key(private val localKeyHandle: LocalKeyHandleKot) {

    companion object {

        fun generate(algorithm: KeyAlgs, ephemeral: Boolean = false, ): Key {
            return Key(Askar.key.keyGenerate(algorithm, ephemeral))
        }

        fun fromSeed(method: KeyMethod = KeyMethod.None, algorithm: KeyAlgs, seed: String): Key {
            return Key(Askar.key.keyFromSeed(algorithm, seed, method))
        }

        fun fromSecretBytes(algorithm: KeyAlgs, secretKey: ByteArray): Key {
            return Key(Askar.key.keyFromSecretBytes(algorithm, secretKey))
        }

        fun fromPublicBytes(algorithm: KeyAlgs, publicKey: ByteArray): Key {
            return Key(Askar.key.keyFromPublicBytes(algorithm, publicKey))
        }

        fun fromJwk(jwk: Jwk): Key {
            return Key(Askar.key.keyFromJwk(jwk))
        }
    }

    fun handle(): LocalKeyHandleKot {
        return this.localKeyHandle
    }

    fun convertKey(algorithm: KeyAlgs): Key {
        return Key(Askar.key.keyConvert(this.localKeyHandle.handle, algorithm))
    }

    fun fromKeyExchange(algorithm: KeyAlgs, publicKey: Key): Key {
        return Key(Askar.key.keyFromKeyExchange(this.localKeyHandle.handle, publicKey.localKeyHandle.handle, algorithm))
    }

    fun algorithm(): KeyAlgs {
        val alg = Askar.key.keyGetAlgorithm(this.localKeyHandle.handle)
        return keyAlgFromString(alg)
    }

    fun ephemeral(): Boolean {
        val num = Askar.key.keyGetEphemeral(this.localKeyHandle.handle)
        return num.toInt() != 0
    }

    fun publicBytes(): ByteArray {
        return Askar.key.keyGetPublicBytes(this.localKeyHandle.handle)
    }

    fun secretBytes(): ByteArray {
        return Askar.key.keyGetSecretBytes(this.localKeyHandle.handle)
    }

    fun jwkPublic(): Jwk {
        return Json.decodeFromString(Askar.key.keyGetJwkPublic(this.localKeyHandle.handle, this.algorithm()))
    }

    fun jwkSecret(): Jwk {
        val buffer = Askar.key.keyGetJwkSecret(this.localKeyHandle.handle)
        return Json.decodeFromString(buffer)
    }

    fun jwkThumbprint(): String {
        return Askar.key.keyGetJwkThumbprint(this.localKeyHandle.handle, this.algorithm())
    }

    fun aeadParams(): askar.AeadParams {
        return Askar.key.keyGetAeadParams(this.localKeyHandle.handle)
    }

    fun aeadRandomNonce(): ByteArray {
        return Askar.key.keyAeadRandomNonce(this.localKeyHandle.handle)
    }

    fun aeadEncrypt(
        message: String,
        nonce: ByteArray = ByteArray(0),
        aad: String = ""
    ): askar.EncryptedBuffer {
        return Askar.key.keyAeadEncrypt(localKeyHandle.handle, message, nonce, aad)
    }

    fun aeadEncrypt(
        message: ByteArray,
        nonce: ByteArray = ByteArray(0),
        aad: ByteArray = byteArrayOf(0)
    ): askar.EncryptedBuffer {
        return Askar.key.keyAeadEncrypt(localKeyHandle.handle, message, nonce, aad)
    }

    fun aeadDecrypt(
        cipherText: ByteArray,
        nonce: ByteArray = ByteArray(0),
        tag: ByteArray = ByteArray(0),
        aad: String = ""
    ): ByteArray {
        return Askar.key.keyAeadDecrypt(localKeyHandle.handle, cipherText, nonce, tag, aad)
    }

    fun signMessage(message: String, sigType: SigAlgs? = null): ByteArray {
        return Askar.key.keySignMessage(this.localKeyHandle.handle, message, sigType)
    }

    fun signMessage(message: ByteArray, sigType: SigAlgs? = null): ByteArray {
        return Askar.key.keySignMessage(this.localKeyHandle.handle, message, sigType)
    }

    fun verifySignature(message: String, signature: ByteArray, sigType: SigAlgs? = null): Boolean {
        val num = Askar.key.keyVerifySignature(this.localKeyHandle.handle, message, signature, sigType)
        return num.toInt() != 0
    }

    fun verifySignature(message: ByteArray, signature: ByteArray, sigType: SigAlgs? = null): Boolean {
        val num = Askar.key.keyVerifySignature(this.localKeyHandle.handle, message, signature, sigType)
        return num.toInt() != 0
    }

    fun wrapKey(other: Key, nonce: String = ""): askar.EncryptedBuffer {
        return Askar.key.keyWrapKey(this.localKeyHandle.handle, other.localKeyHandle.handle, nonce)
    }

    fun unwrapKey(algorithm: KeyAlgs, tag: String = "", cipherText: ByteArray = ByteArray(0), nonce: ByteArray = byteArrayOf()): Key {
        return Key(Askar.key.keyUnwrapKey(this.localKeyHandle.handle, algorithm, cipherText, nonce, tag))
    }

}

File: wrappers/javascript/packages/aries-askar-shared/src/crypto/Key.ts
import type { LocalKeyHandle } from './handles'
import type { KeyAlg, SigAlgs } from '../enums'

import { Buffer } from 'buffer'

import { ariesAskar } from '../ariesAskar'
import { KeyMethod, keyAlgFromString } from '../enums'

import { Jwk } from './Jwk'

export class Key {
  private localKeyHandle: LocalKeyHandle

  public constructor(handle: LocalKeyHandle) {
    this.localKeyHandle = handle
  }

  public static generate(algorithm: KeyAlg, ephemeral = false) {
    return new Key(ariesAskar.keyGenerate({ algorithm, ephemeral }))
  }

  public static fromSeed({
    method = KeyMethod.None,
    algorithm,
    seed,
  }: {
    algorithm: KeyAlg
    seed: Uint8Array
    method?: KeyMethod
  }) {
    return new Key(ariesAskar.keyFromSeed({ algorithm, method, seed }))
  }

  public static fromSecretBytes(options: { algorithm: KeyAlg; secretKey: Uint8Array }) {
    return new Key(ariesAskar.keyFromSecretBytes(options))
  }

  public static fromPublicBytes(options: { algorithm: KeyAlg; publicKey: Uint8Array }) {
    return new Key(ariesAskar.keyFromPublicBytes(options))
  }

  public static fromJwk(options: { jwk: Jwk }) {
    return new Key(ariesAskar.keyFromJwk(options))
  }

  public convertkey(options: { algorithm: KeyAlg }) {
    return new Key(ariesAskar.keyConvert({ localKeyHandle: this.handle, ...options }))
  }

  public keyFromKeyExchange({ algorithm, publicKey }: { algorithm: KeyAlg; publicKey: Key }) {
    return new Key(ariesAskar.keyFromKeyExchange({ skHandle: this.handle, pkHandle: publicKey.handle, algorithm }))
  }

  public get handle() {
    return this.localKeyHandle
  }

  public get algorithm() {
    const alg = ariesAskar.keyGetAlgorithm({ localKeyHandle: this.handle })
    return keyAlgFromString(alg)
  }

  public get ephemeral() {
    return Boolean(ariesAskar.keyGetEphemeral({ localKeyHandle: this.handle }))
  }

  public get publicBytes() {
    return ariesAskar.keyGetPublicBytes({ localKeyHandle: this.handle })
  }

  public get secretBytes() {
    return ariesAskar.keyGetSecretBytes({ localKeyHandle: this.handle })
  }

  public get jwkPublic(): Jwk {
    return Jwk.fromString(ariesAskar.keyGetJwkPublic({ localKeyHandle: this.handle, algorithm: this.algorithm }))
  }

  public get jwkSecret() {
    const secretBytes = ariesAskar.keyGetJwkSecret({ localKeyHandle: this.handle })
    return Jwk.fromString(Buffer.from(secretBytes).toString())
  }

  public get jwkThumbprint() {
    return ariesAskar.keyGetJwkThumbprint({ localKeyHandle: this.handle, algorithm: this.algorithm })
  }

  public get aeadParams() {
    return ariesAskar.keyAeadGetParams({ localKeyHandle: this.handle })
  }

  public get aeadRandomNonce() {
    return ariesAskar.keyAeadRandomNonce({ localKeyHandle: this.handle })
  }

  public aeadEncrypt(options: { message: Uint8Array; nonce?: Uint8Array; aad?: Uint8Array }) {
    return ariesAskar.keyAeadEncrypt({ localKeyHandle: this.handle, ...options })
  }

  public aeadDecrypt(options: { ciphertext: Uint8Array; nonce: Uint8Array; tag?: Uint8Array; aad?: Uint8Array }) {
    return ariesAskar.keyAeadDecrypt({ localKeyHandle: this.handle, ...options })
  }

  public signMessage(options: { message: Uint8Array; sigType?: SigAlgs }) {
    return ariesAskar.keySignMessage({ localKeyHandle: this.handle, ...options })
  }

  public verifySignature(options: { message: Uint8Array; signature: Uint8Array; sigType?: SigAlgs }) {
    return ariesAskar.keyVerifySignature({ localKeyHandle: this.handle, ...options })
  }

  public wrapKey({ other, nonce }: { other: Key; nonce?: Uint8Array }) {
    return ariesAskar.keyWrapKey({ localKeyHandle: this.handle, other: other.handle, nonce })
  }

  public unwrapKey(options: { algorithm: KeyAlg; tag?: Uint8Array; ciphertext: Uint8Array; nonce?: Uint8Array }) {
    return new Key(ariesAskar.keyUnwrapKey({ localKeyHandle: this.handle, ...options }))
  }
}
